File: internal/parsing/system_event.go
package parsing

import (
	"encoding/json"

	"claude-devtools/internal/domain"
)

// buildSystemEventData mirrors system_event::build_system_event_data — typed
// payloads for the four displayable system subtypes (others → nil).
func buildSystemEventData(e *domain.RawJsonlEntry) *domain.SystemEventData {
	if e.Subtype == nil {
		return nil
	}
	subtype := *e.Subtype
	switch subtype {
	case "api_error":
		var errorStatus *uint16
		var errorType, errorMessage *string
		if len(e.Error) > 0 {
			var em map[string]json.RawMessage
			if json.Unmarshal(e.Error, &em) == nil {
				if v := asU64opt(em["status"]); v != nil {
					s := uint16(*v)
					errorStatus = &s
				}
				if inner, ok := em["error"]; ok {
					var im map[string]json.RawMessage
					if json.Unmarshal(inner, &im) == nil {
						if t, ok := decodeString(im["type"]); ok {
							errorType = &t
						}
						if msg, ok := decodeString(im["message"]); ok {
							errorMessage = &msg
						}
					}
				}
			}
		}
		if errorType == nil && len(e.Cause) > 0 {
			var cm map[string]json.RawMessage
			if json.Unmarshal(e.Cause, &cm) == nil {
				if code, ok := decodeString(cm["code"]); ok {
					errorType = &code
				}
			}
		}
		return &domain.SystemEventData{
			Subtype:      subtype,
			ErrorStatus:  errorStatus,
			ErrorType:    errorType,
			ErrorMessage: errorMessage,
			RetryAttempt: e.RetryAttempt,
			MaxRetries:   e.MaxRetries,
			RetryInMs:    e.RetryInMs,
		}
	case "bridge_status":
		return &domain.SystemEventData{Subtype: subtype, BridgeContent: e.Content, BridgeURL: e.URL}
	case "memory_saved":
		return &domain.SystemEventData{Subtype: subtype, WrittenPaths: e.WrittenPaths, MemoryVerb: e.Verb}
	case "turn_duration":
		return &domain.SystemEventData{Subtype: subtype, DurationMs: e.DurationMs}
	default:
		return nil
	}
}

File: src-tauri/src/parsing/system_event.rs
use crate::types::jsonl::RawJsonlEntry;
use crate::types::messages::SystemEventData;

pub fn build_system_event_data(entry: &RawJsonlEntry) -> Option<SystemEventData> {
    let subtype = entry.subtype.as_deref()?;
    match subtype {
        "api_error" => {
            let mut error_status: Option<u16> = None;
            let mut error_type: Option<String> = None;
            let mut error_message: Option<String> = None;

            if let Some(ref err) = entry.error {
                error_status = err
                    .get("status")
                    .and_then(|v| v.as_u64())
                    .map(|v| v as u16);
                if let Some(inner) = err.get("error") {
                    error_type = inner
                        .get("type")
                        .and_then(|v| v.as_str())
                        .map(|s| s.to_string());
                    error_message = inner
                        .get("message")
                        .and_then(|v| v.as_str())
                        .map(|s| s.to_string());
                }
            }
            if error_type.is_none() {
                if let Some(ref cause) = entry.cause {
                    error_type = cause
                        .get("code")
                        .and_then(|v| v.as_str())
                        .map(|s| s.to_string());
                }
            }

            Some(SystemEventData {
                subtype: subtype.to_string(),
                error_status,
                error_type,
                error_message,
                retry_attempt: entry.retry_attempt,
                max_retries: entry.max_retries,
                retry_in_ms: entry.retry_in_ms,
                ..Default::default()
            })
        }
        "bridge_status" => Some(SystemEventData {
            subtype: subtype.to_string(),
            bridge_content: entry.content.clone(),
            bridge_url: entry.url.clone(),
            ..Default::default()
        }),
        "memory_saved" => Some(SystemEventData {
            subtype: subtype.to_string(),
            written_paths: entry.written_paths.clone(),
            memory_verb: entry.verb.clone(),
            ..Default::default()
        }),
        "turn_duration" => Some(SystemEventData {
            subtype: subtype.to_string(),
            duration_ms: entry.duration_ms,
            ..Default::default()
        }),
        _ => None,
    }
}
